### PR TITLE
Try less memory usage

### PR DESCRIPTION
**Pull request by MBkkt**
0.4 vs 0.5 on my machine with old memset.
And 0.1 without unnecessary memset

### CODE REVIEW

File: core/formats/columnstore2.hpp [hunk 1]
```diff
@@ -82,45 +82,46 @@ class column final : public irs::column_output {
 
   class address_table {
    public:
+    address_table(ManagedTypedAllocator<uint64_t> alloc) : alloc_{alloc} {
+      offsets_ = alloc_.allocate(kBlockSize);
```
Comment (MBkkt) [inline core/formats/columnstore2.hpp:86]: Now we allocate very good aligned blocks, so I think it's better utilize allocator.

File: core/formats/columnstore2.hpp [hunk 2]
```diff
@@ -82,45 +82,46 @@ class column final : public irs::column_output {
...
       IRS_ASSERT(offset_ < offsets_ + kBlockSize);
       *offset_++ = offset;
-      IRS_ASSERT(offset >= offset_[-1]);
     }
 
     void pop_back() noexcept {
-      IRS_ASSERT(offset_ > offsets_);
-      *--offset_ = 0;
+      IRS_ASSERT(offsets_ < offset_);
+      --offset_;
```
Comment (MBkkt) [inline core/formats/columnstore2.hpp:104]: we don't need to set zero to unused data

File: core/formats/columnstore2.hpp [hunk 3]
```diff
@@ -82,45 +82,46 @@ class column final : public irs::column_output {
...
+      return static_cast<uint32_t>(offset_ - offsets_);
     }
 
     bool empty() const noexcept { return offset_ == offsets_; }
 
-    bool full() const noexcept { return offset_ == std::end(offsets_); }
+    bool full() const noexcept { return offset_ == offsets_ + kBlockSize; }
 
-    void reset() noexcept {
-      std::memset(offsets_, 0, sizeof offsets_);
```
Comment (MBkkt) [inline core/formats/columnstore2.hpp:113]: we don't need to set zero to unused data

File: core/formats/columnstore2.hpp [hunk 4]
```diff
@@ -82,45 +82,46 @@ class column final : public irs::column_output {
...
+    void reset() noexcept { offset_ = offsets_; }
 
-    uint64_t* begin() noexcept { return std::begin(offsets_); }
+    uint64_t* begin() noexcept { return offsets_; }
     uint64_t* current() noexcept { return offset_; }
-    uint64_t* end() noexcept { return std::end(offsets_); }
+    uint64_t* end() noexcept { return offsets_ + kBlockSize; }
 
    private:
-    uint64_t offsets_[kBlockSize]{};
```
Comment (MBkkt) [inline core/formats/columnstore2.hpp:122]: We don't need zero unused data

File: core/formats/columnstore2.hpp [hunk 5]
```diff
@@ -82,45 +82,46 @@ class column final : public irs::column_output {
 
   class address_table {
    public:
+    address_table(ManagedTypedAllocator<uint64_t> alloc) : alloc_{alloc} {
+      offsets_ = alloc_.allocate(kBlockSize);
+      offset_ = offsets_;
+    }
+
+    ~address_table() { alloc_.deallocate(offsets_, kBlockSize); }
```
Comment (MBkkt) [inline core/formats/columnstore2.hpp:90]: Also deque with columns works better now:
blocks in deque could handle few columns and we don't store this memory forever.
only while write segment

File: core/formats/columnstore2.hpp (hunk 1 above)
Comment (Dronplane) [reply core/formats/columnstore2.hpp:86]: Here would be 2 allocations. One for deque itself and this one for internal contents. Better to benchmark this on test load with many columns.

File: microbench/CMakeLists.txt [hunk 6]
```diff
@@ -76,3 +76,21 @@ target_include_directories(iresearch-memory-file-bench
 target_link_libraries(iresearch-memory-file-bench
   iresearch-static
   )
+
+add_executable(iresearch-column-writer
+  column_writer_benchmark.cpp
```
Comment (Dronplane) [inline microbench/CMakeLists.txt:81]: Missing file?

File: microbench/column_writer_benchmark.cpp [hunk 7]
```diff
@@ -0,0 +1,35 @@
+#include <random>
```
Comment (Dronplane) [inline microbench/column_writer_benchmark.cpp:1]: Disclaimer missing

**Review by Dronplane — APPROVED** [review]
LGTM

**Pull request merged**